// @bissolli/storybook-css-properties/dist/esm/Panel.js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]); if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

import React, { useState, useEffect } from "react";
import { AddonPanel } from "@storybook/components";
import { PanelContent } from "./components/PanelContent";
import { getAllCSSVariables } from "./get-all-css-variables";
import { ADDON_ID, PARAM_KEY } from "./constants";
import { useParameter } from "@storybook/api";
import { useLocalStorage } from "./use-local-storage";

var setIframePreviewWhenReady = function setIframePreviewWhenReady(setIframePreview) {
  var iframePreview = document.getElementById('storybook-preview-iframe');

  if (!iframePreview) {
    setTimeout(function () {
      return setIframePreviewWhenReady(setIframePreview);
    }, 2000);
    return;
  }

  setIframePreview(iframePreview);
};

export var Panel = function Panel(props) {
  var _paramData$props, _paramData$matchCateg, _paramData$hiddenProp;

  var storage = useLocalStorage();
  var paramData = useParameter(PARAM_KEY, {});

  var _useState = useState([]),
      _useState2 = _slicedToArray(_useState, 2),
      cssVars = _useState2[0],
      setCssVars = _useState2[1];

  var _useState3 = useState(null),
      _useState4 = _slicedToArray(_useState3, 2),
      iframePreview = _useState4[0],
      setIframePreview = _useState4[1];

  setTimeout(function () {
    return setIframePreviewWhenReady(setIframePreview);
  }, 2000);
  useEffect(function () {
    if (!iframePreview) return;
    var variables = getAllCSSVariables(iframePreview.contentWindow.document);
    setCssVars(variables);
    storage.setLocalStorage(ADDON_ID, {
      initialValues: variables
    });
  }, [iframePreview]);
  return /*#__PURE__*/React.createElement(AddonPanel, props, /*#__PURE__*/React.createElement(PanelContent, {
    baseProperties: cssVars,
    propsConfig: (_paramData$props = paramData === null || paramData === void 0 ? void 0 : paramData.props) !== null && _paramData$props !== void 0 ? _paramData$props : {},
    matchCategory: (_paramData$matchCateg = paramData === null || paramData === void 0 ? void 0 : paramData.matchCategory) !== null && _paramData$matchCateg !== void 0 ? _paramData$matchCateg : {},
    hiddenProps: (_paramData$hiddenProp = paramData === null || paramData === void 0 ? void 0 : paramData.hiddenProps) !== null && _paramData$hiddenProp !== void 0 ? _paramData$hiddenProp : []
  }));
};
