// @bissolli/storybook-css-properties/dist/esm/use-local-storage.js
var store = {};

var virtualStorage = function virtualStorage() {
  return {
    getItem: function getItem(key) {
      return store[key] || null;
    },
    setItem: function setItem(key, value) {
      store[key] = value.toString();
    },
    removeItem: function removeItem(key) {
      delete store[key];
    },
    clear: function clear() {
      store = {};
    },
    key: function key(_) {
      return null;
    },
    length: Object.keys(store).length
  };
};

var isAvailable = function isAvailable() {
  var test = 'test';

  try {
    window.localStorage.setItem(test, test);
    window.localStorage.removeItem(test);
    return true;
  } catch (e) {
    return false;
  }
};

var getLocalStorage = function getLocalStorage() {
  return isAvailable() ? window.localStorage : virtualStorage();
};

var decodeValue = function decodeValue(value) {
  if (typeof value === 'string') {
    try {
      return JSON.parse(value);
    } catch (_) {//
    }
  }

  return value;
};

var encodeValue = function encodeValue(val) {
  if (typeof val === 'string') {
    return val;
  }

  return JSON.stringify(val);
};

export var useLocalStorage = function useLocalStorage() {
  var localStorage = getLocalStorage();

  var getItem = function getItem(key) {
    var value = localStorage.getItem(key);
    return decodeValue(value);
  };

  var setItem = function setItem(key, value) {
    return localStorage.setItem(key, encodeValue(value));
  };

  var removeItem = function removeItem(key) {
    return localStorage.removeItem(key);
  };

  return {
    decodeValue: decodeValue,
    encodeValue: encodeValue,
    getLocalStorage: getItem,
    setLocalStorage: setItem,
    removeLocalStorage: removeItem
  };
};
